crates/sparrow-journal/src/checkpoints.rs

```diff
@@ -0,0 +1,254 @@
+fn recover_checkpoints(
+    checkpoint_path: &std::path::Path,
+    schema: &SchemaRef,
+) -> error_stack::Result<Vec<CheckpointFile>, Error> {
+    let mut checkpoints = vec![];
+    if checkpoint_path.exists() {
+        error_stack::ensure!(
+            checkpoint_path.is_dir(),
+            Error::CheckpointDirNotDir(checkpoint_path.to_path_buf())
+        );
+    } else {
+        std::fs::create_dir_all(checkpoint_path)
+            .into_report()
+            .change_context(Error::RecoveringCheckpoints)?;
+    }
+
+    for file in std::fs::read_dir(checkpoint_path)
+        .into_report()
+        .change_context(Error::RecoveringCheckpoints)?
+    {
+        let file = file
+            .into_report()
+            .change_context(Error::RecoveringCheckpoints)?;
+        let file_type = file
+            .file_type()
+            .into_report()
+            .change_context(Error::RecoveringCheckpoints)?;
+        if !file_type.is_file() || file.path().extension() != Some("arrow".as_ref()) {
+            continue;
+        }
+
+        let reader = std::fs::File::open(file.path())
+            .into_report()
+            .change_context(Error::RecoveringCheckpoints)?;
+        let reader = arrow_ipc::reader::FileReader::try_new(reader, None)
+            .into_report()
+            .change_context(Error::RecoveringCheckpoints)?;
+        error_stack::ensure!(&reader.schema() == schema, Error::RecoveringCheckpoints);
+
+        let file_name = file
+            .file_name()
+            .into_string()
+            .map_err(|_| Error::RecoveringCheckpoints)?;
+
+        let min_present_time = reader
+            .custom_metadata()
+            .get("min_present_time")
+            .ok_or(Error::RecoveringCheckpoints)?;
+        let min_present_time =
+            RowTime::from_str(min_present_time).change_context(Error::RecoveringCheckpoints)?;
+
+        let max_present_time = reader
+            .custom_metadata()
+            .get("max_present_time")
+            .ok_or(Error::RecoveringCheckpoints)?;
+        let max_present_time =
+            RowTime::from_str(max_present_time).change_context(Error::RecoveringCheckpoints)?;
+
+        checkpoints.push(CheckpointFile {
+            file_name,
+            min_present_time,
+            max_present_time,
+        })
+    }
+
+    Ok(checkpoints)
+}
+
+impl Checkpoints {
+    pub fn try_new(checkpoint_dir: PathBuf, schema: SchemaRef) -> error_stack::Result<Self, Error> {
+        let checkpoints = recover_checkpoints(&checkpoint_dir, &schema)?;
+        tracing::info!(
+            "Recovered {} checkpoints from {}",
+            checkpoints.len(),
+            checkpoint_dir.display()
+        );
+        Ok(Self {
+            checkpoint_dir,
+            checkpoints,
+            schema,
+        })
+    }
+
+    /// Create a check-point from the given batches.
+    ///
+    /// This method assumes the batches are non-overlapping and ordered by time.
+    pub fn checkpoint(
```

Comment:
At this time, journals will include the "prepared" data, correct? (we have the Batch with separate time,subsort,key columns, so I assume that must be the case) Is there any reason we'd want to instead journal the user data directly? For example, if we support decimals in `prepare`, or any non-backwards-compatible change, we'd want to run prepare on the user data again.